File: src/openjarvis/mcp/protocol.py
"""MCP JSON-RPC 2.0 protocol message types."""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any, Dict, Optional

# Error codes per JSON-RPC 2.0 / MCP spec
PARSE_ERROR = -32700
INVALID_REQUEST = -32600
METHOD_NOT_FOUND = -32601
INVALID_PARAMS = -32602
INTERNAL_ERROR = -32603


@dataclass
class MCPRequest:
    """JSON-RPC 2.0 request message."""

    method: str
    params: Dict[str, Any] = field(default_factory=dict)
    id: int | str = 0
    jsonrpc: str = "2.0"

    def to_json(self) -> str:
        """Serialize to JSON string."""
        return json.dumps(
            {
                "jsonrpc": self.jsonrpc,
                "id": self.id,
                "method": self.method,
                "params": self.params,
            }
        )

    @classmethod
    def from_json(cls, data: str) -> MCPRequest:
        """Deserialize from JSON string."""
        parsed = json.loads(data)
        return cls(
            method=parsed["method"],
            params=parsed.get("params", {}),
            id=parsed.get("id", 0),
            jsonrpc=parsed.get("jsonrpc", "2.0"),
        )


@dataclass
class MCPResponse:
    """JSON-RPC 2.0 response message."""

    result: Any = None
    error: Optional[Dict[str, Any]] = None
    id: int | str = 0
    jsonrpc: str = "2.0"

    def to_json(self) -> str:
        """Serialize to JSON string."""
        obj: Dict[str, Any] = {"jsonrpc": self.jsonrpc, "id": self.id}
        if self.error is not None:
            obj["error"] = self.error
        else:
            obj["result"] = self.result
        return json.dumps(obj)

    @classmethod
    def from_json(cls, data: str) -> MCPResponse:
        """Deserialize from JSON string."""
        parsed = json.loads(data)
        return cls(
            result=parsed.get("result"),
            error=parsed.get("error"),
            id=parsed.get("id", 0),
            jsonrpc=parsed.get("jsonrpc", "2.0"),
        )

    @classmethod
    def error_response(
        cls,
        id: int | str,
        code: int,
        message: str,
        data: Any = None,
    ) -> MCPResponse:
        """Create an error response."""
        error: Dict[str, Any] = {"code": code, "message": message}
        if data is not None:
            error["data"] = data
        return cls(error=error, id=id)


@dataclass
class MCPNotification:
    """JSON-RPC 2.0 notification (no id, no response expected)."""

    method: str
    params: Dict[str, Any] = field(default_factory=dict)
    jsonrpc: str = "2.0"

    def to_json(self) -> str:
        """Serialize to JSON string."""
        return json.dumps(
            {
                "jsonrpc": self.jsonrpc,
                "method": self.method,
                "params": self.params,
            }
        )


@dataclass
class MCPError(Exception):
    """MCP protocol error with JSON-RPC error code."""

    code: int
    message: str
    data: Any = None

    def __str__(self) -> str:
        return f"MCPError({self.code}): {self.message}"


__all__ = [
    "INTERNAL_ERROR",
    "INVALID_PARAMS",
    "INVALID_REQUEST",
    "MCPError",
    "MCPNotification",
    "MCPRequest",
    "MCPResponse",
    "METHOD_NOT_FOUND",
    "PARSE_ERROR",
]

File: repos/openjarvis/rust/crates/openjarvis-mcp/src/protocol.rs
//! MCP JSON-RPC 2.0 protocol types.

use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct McpRequest {
    pub jsonrpc: String,
    pub method: String,
    #[serde(default)]
    pub params: Value,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<Value>,
}

impl McpRequest {
    pub fn new(method: &str, params: Value, id: Value) -> Self {
        Self {
            jsonrpc: "2.0".into(),
            method: method.to_string(),
            params,
            id: Some(id),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct McpResponse {
    pub jsonrpc: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub result: Option<Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<McpError>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct McpError {
    pub code: i64,
    pub message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data: Option<Value>,
}

impl McpResponse {
    pub fn success(id: Value, result: Value) -> Self {
        Self {
            jsonrpc: "2.0".into(),
            result: Some(result),
            error: None,
            id: Some(id),
        }
    }

    pub fn error(id: Value, code: i64, message: &str) -> Self {
        Self {
            jsonrpc: "2.0".into(),
            result: None,
            error: Some(McpError {
                code,
                message: message.to_string(),
                data: None,
            }),
            id: Some(id),
        }
    }

    pub fn method_not_found(id: Value) -> Self {
        Self::error(id, -32601, "Method not found")
    }

    pub fn invalid_params(id: Value, msg: &str) -> Self {
        Self::error(id, -32602, msg)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_request_serde() {
        let req = McpRequest::new(
            "tools/list",
            serde_json::json!({}),
            serde_json::json!(1),
        );
        let json = serde_json::to_string(&req).unwrap();
        let parsed: McpRequest = serde_json::from_str(&json).unwrap();
        assert_eq!(parsed.method, "tools/list");
    }

    #[test]
    fn test_response_success() {
        let resp = McpResponse::success(
            serde_json::json!(1),
            serde_json::json!({"tools": []}),
        );
        assert!(resp.result.is_some());
        assert!(resp.error.is_none());
    }

    #[test]
    fn test_response_error() {
        let resp = McpResponse::method_not_found(serde_json::json!(1));
        assert!(resp.error.is_some());
        assert_eq!(resp.error.unwrap().code, -32601);
    }
}
